### PR TITLE
Representation differentials -- building blocks for proper motion and radial velocity

### PR DESCRIPTION
**Pull request by mhvk**
@eteq, @adrn, @astrofrog -- following our discussion about proper motions and radial velocities, I went ahead an implemented the low-level building blocks: the ability to deal with offsets and derivatives in different representations. It definitely required a lot of mind-warping, but I'm fairly sure what I have here is roughly what we need. To get a sense of how it works, maybe start with just looking at the short addition to the documentation (and I'd recommend to look at the test cases before turning to the code).

For the code itself, I ended up doing quite a refactor, so the changes may be larger than you'd really want; it may help to go in order of the commits...

### CODE REVIEW

**Comment by mhvk** [comment]
Improved docstrings (i do assume this will get in in some form or another).

Also, realised it would be quite easy to make a `SphericalCosLatDifferential` class to support proper motions with the `cos(dec)` term (which I think we need). But I'll await comments before doing that; an alternative would be an argument that tells whether or not the `cos` is included.

**Comment by eteq** [comment]
Wow, didn't expect you to run off with this @mhvk - nice!  @adrn and I are planning to hack on the velocities stuff in ~2 weeks, so I expect we'll have a much closer look then...

**Comment by adrn** [comment]
@mhvk As another trial of this machinery, I've been experimenting with making use of the differential classes in the dynamics code I have. The main place where these classes are a natural fit is in my own class for representing phase-space positions (positions and velocities/momenta) – see commits above if interested. While playing around, I've been jotting down things that would be nice to have:

* Add a `d_xyz` attribute to `CartesianDifferential`, like `.xyz` in the representation
* Provide an easy way (even just a dictionary) to get the corresponding Differential class given a representation class, and vice versa. 
* It looks like cross-products between representations and differentials don't work? Was this intentional or is this something we can support? It would be very useful for computing, e.g., angular momentum vectors.
* Have you thought at all above providing functions like `numpy.stack` or `numpy.concatenate` for the representation classes?

Sorry that @eteq and I are being slow about spec'ing out the velocity transforms in astropy core...hopefully we get to that soon.

**Comment by mhvk** [comment]
@adrn - Happy to hear it is starting to be useful! On the four ideas:
* `d_xyz` should be easy and makes sense.
* The lookup of the corresponding base class does exist: `differential.base_representation`; I had thought about the reverse, but wasn't quite sure what was the best way, and wanted to be sure before doing anything that started to affect the representations themselves. E.g., one could have a direct lookup, or think more in terms of a method that, e.g., returns a difference as a differential. In what situation do you need to use it? Simply a transformation of rep+diff to another system? I ask in part since I did wonder about having a container that holds both...
* The cross products should work! For this case, I think it would be reasonable to assume that the "base" of the differential is the representation, right? 
* Yes, definitely have wondered about stacking, also more generally for `Time`, `SkyCoord`, etc. It would have been quite simple if those numpy functions allowed overrides or even just worked well with ndarray subclasses, but sadly they do not, so it needs to start with doing it right on `Quantity`. @taldcroft's work on making an `empty_like` protocol in #5841 may be a good start for this (which reminds me I should have a closer look at that...).

And don't worry about speed -- I don't have much time to work on this either at the moment...

**Comment by mhvk** [comment]
@adrn - I'll answer here to your https://github.com/mhvk/astropy/pull/24, to keep the discussion in one place.

On the move of `get_name`: I had left it on the representation initially as I wanted to ensure we didn't end up mixing in differentials into `REPRESENTATION_BASES`, but later split the metaclass, which makes this impossible anyway. So, I think this change is fine. I would only suggest adding a few test cases.

For the `d_xyz` property, that looks fine, and my only comment is that since the code is almost identical to that of `Cartesian.xyz` it may make sense to use a utility function that takes `x, y, z, xyz_axis` and turns it into an array (replacing `self.ndim` and `self.shape` with `x.ndim` and `x.shape`). Also for this, probably good to add a test.

Anyway, good additions! More generally, are the differentials holding up to their potential?

**Comment by adrn** [comment]
@mhvk For the `get_name` tests, are you imagining just checking that, e.g., `SphericalRepresentation.get_name() == 'spherical'`, `SphericalDifferential.get_name() == 'spherical'`, etc.?

Yes, the differentials do seem quite useful! I'm still working on integrating them into my own code but I haven't hit any showstoppers yet.

**Comment by mhvk** [comment]
Yes, as simple as that!

**Comment by adrn** [comment]
Sounds good! Pushed.

**Comment by adrn** [comment]
@mhvk Actually, back to the topic of `get_name()` and `REPRESENTATION_CLASSES`, how would you feel about having a `DIFFERENTIAL_CLASSES` dict as well?

**Comment by mhvk** [comment]
I'm all-OK with having a `DIFFERENTIAL_CLASSES` dict -- you're the user right now and are finding out what is needed! Indeed, in the end we may want a more direct way, and perhaps just have an attribute on the base class (or have combined classes). The main reason to hold off on actually adding stuff to the representations themselves would be that, e.g., at some point we might want to add a differential that uses `d_lon_coslat, d_lat, d_distance` and if we do that, it becomes a bit unclear how to indicate what differential belongs to which representation.

Anyway, bottom line: do go ahead and add a dict for the differential classes; it seems quite clear that that will be useful no matter what.

**Comment by mhvk** [comment]
@adrn - do you want to do the `DIFFERENTIAL_CLASSES` in the same PR, or shall I just merge the current one (which I think is all OK)?

**Comment by adrn** [comment]
I'll add it in the existing PR and push shortly.

**Comment by adrn** [comment]
@mhvk OK, I added `DIFFERENTIAL_CLASSES` and some very simple tests.

**Comment by adrn** [comment]
@mhvk You'll be pleased to know that (here at pyastro17) @eteq and I each got different implementations of velocity transforms working using the coordinate transformation framework! They are just hacks at the moment, but I implemented the `AffineTransform` / pure linear algebraic transform scheme for ICRS->Galactocentric, and @eteq got the numerical differentiation scheme working (which will be needed for some transformations that can't be expressed as matrix operations). I'm currently working on a document that explains the main ideas, a path forward, and which raises a few questions about this representation differential machinery that you might have answers to. Will link to that soon!

**Comment by adrn** [comment]
@mhvk Here are some notes. Feel free to add to this @eteq! https://docs.google.com/document/d/1dZIG7CG8h-C4lcegtx71eUc76K5LHtgRBsXk4iRa2ds/edit#heading=h.2qib7gau1ysm

**Comment by adrn** [comment]
@mhvk @eteq I should also note that I've started implementing some things that are mentioned in the doc above in a local branch (mainly the `AffineTransform`). But since it is based off of this PR, I'm not sure what the best workflow is for working collaboratively on that and then later submitting a PR to astropy/master?

**Comment by mhvk** [comment]
@adrn - you can always make a PR with this PR included. But, if it seems that this is usable and at least in general the way to go, maybe we should just finish review and merge this? If need be, we can still make changes, at least before 2.0. (Here, of course, your experience actually using it is most relevant !)

**Comment by mhvk** [comment]
@adrn, @eteq - I made some edits to the google doc (search for MHvK). For this PR, it seems the main to-do item would be to introduce a spherical differential that includes coslat, correct?

**Comment by adrn** [comment]
@mhvk I think that's right!

**Comment by mhvk** [comment]
@adrn, @eteq - I implemented classes that have the `coslat` term included. Please have a look (documentation slightly sparse, but possibly the best place to start nevertheless).

**Comment by mhvk** [comment]
(test failure is a false positive - URL timeout)

File: docs/coordinates/representations.rst [hunk 1]
```diff
@@ -212,18 +212,123 @@ To see how the operations work, consider the following examples::
         (  3.06161700e-16,  -5.00000000e+00,  12.),
         ( -4.00000000e+00,  -3.00000000e+00, -12.)]]>
 
+For coordinates, one also has to deal with proper motions, which describe
```
Comment (eteq) [inline docs/coordinates/representations.rst]: (As with the two docstrings above, I pushed up some changes to this section that I think are controversial.)

File: astropy/coordinates/representation.py [hunk 2]
```diff
@@ -20,23 +20,29 @@
...
+           "SphericalDifferential", "SphericalCosLatDifferential",
+           "UnitSphericalDifferential", "UnitSphericalCosLatDifferential",
+           "RadialDifferential", "CylindricalDifferential",
+           "PhysicsSphericalDifferential"]
+
+# Module-level dict mapping representation string alias names to classes.
+# This is populated by the metaclass init so all representation and differential
+# classes get registered automatically.
 REPRESENTATION_CLASSES = {}
+DIFFERENTIAL_CLASSES = {}
```
Comment (adrn) [inline astropy/coordinates/representation.py]: In the current formulation, these have to be separate because the dict's are keyed by the short name. i.e. `REPRESENTATION_CLASSES['spherical']` and `DIFFERENTIAL_CLASSES['spherical']` have the same key name, but return different classes.

File: astropy/coordinates/representation.py [hunk 3]
```diff
@@ -744,15 +881,42 @@ def lat(self):
         """
         return self._lat
 
+    def unit_vectors(self):
```
Comment (adrn) [inline astropy/coordinates/representation.py]: No good ideas here. Something about "basis"?

File: astropy/coordinates/representation.py [hunk 4]
```diff
@@ -59,68 +65,112 @@ def fmt(x):
...
+    Parameters
+    ----------
+    comp1, comp2, comp3 : `~astropy.units.Quantity` or subclass
+        The components of the 3D point or differential.  The names are the
+        keys and the subclasses the values of the ``attr_classes`` attribute.
+    copy : bool, optional
+        If `True` (default), arrays will be copied rather than referenced.
     """
 
-    recommended_units = {}  # subclasses can override
```
Comment (adrn) [inline astropy/coordinates/representation.py]: Should this go on the super super class (`RepresentationBase` in the current iteration)? That is, do we want to allow `Differential` classes to have `recommended_units`?

File: astropy/coordinates/representation.py [hunk 5]
```diff
@@ -20,23 +20,29 @@
...
-__all__ = ["BaseRepresentation", "CartesianRepresentation",
-           "SphericalRepresentation", "UnitSphericalRepresentation",
-           "PhysicsSphericalRepresentation", "CylindricalRepresentation"]
-
-# Module-level dict mapping representation string alias names to class.
-# This is populated by the metaclass init so all representation classes
-# get registered automatically.
+from ..utils.compat.numpy import broadcast_arrays, broadcast_to
+
+__all__ = ["RepresentationBase", "BaseRepresentation",
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: OK, the name is indeed bad. Not too fond of the very long name, but cannot think of anything better, so let's just go with it.

File: astropy/coordinates/representation.py (hunk 2 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: Indeed, in the current implementation, they have to be separate.

File: astropy/coordinates/representation.py [hunk 6]
```diff
@@ -444,54 +625,42 @@ def __init__(self, x, y=None, z=None, unit=None, xyz_axis=None, copy=True):
...
-            raise ValueError("Input parameters x, y, and z cannot be broadcast")
+        super(CartesianRepresentation, self).__init__(x, y, z, copy=copy)
+        if not (self._x.unit.physical_type ==
+                self._y.unit.physical_type == self._z.unit.physical_type):
+            raise u.UnitsError("x, y, and z should have matching physical types")
 
-        self._x = x
-        self._y = y
-        self._z = z
+    def unit_vectors(self):
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: I went intermediate: I added them to `BaseRepresentation`, but by default raising `NotImplementedError` -- I don't think numerical differentiation should be done by default; if we go that route, we should define a mixin that can be used.

File: astropy/coordinates/representation.py (hunk 3 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: I think I'll let this stand -- in the context of representations, unit vectors is really the logical term...

File: astropy/coordinates/representation.py (hunk 4 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: I'd actually like to remove this completely in a follow-up PR -- these recommended units are set but not used anywhere in the representations proper; this really belongs in the coordinate classes -- where presently they seem to be overridden.

But for differentials I see even less reason to have a recommended unit.

Comment (adrn) [inline astropy/coordinates/representation.py]: 👍  Cool, agreed - I mainly wanted them for the frames, but that does seem like the more logical place for them.

**Comment by adrn** [comment]
Something I just noticed (not because of this PR): in `test_representations.py`, under the `Cartesian` test class, `test_init_singleunit()` is the same as `test_init_quantity()`.

**Comment by mhvk** [comment]
@adrn - OK, fixed that while I was at it.

**Comment by adrn** [comment]
Tests are passing! 🎉

I think @eteq and I should have one last look at this, then merge.

File: astropy/coordinates/representation.py (hunk 6 above)
Comment (eteq) [inline astropy/coordinates/representation.py]: Sure, that's fine.  I want to argue and say we *should* do it, but I realize this is really an edge case and probably isn't worth the effort ;)

**Comment by bsipocz** [comment]
This is still missing a changelog, so please add one before merging.

**Comment by eteq** [comment]
Looks all good now - I'll update the changelog manually and then merge.  Thanks @mhvk ! (And thanks @bsipocz  for being a "pre-@astrobot" :wink: )
